### ISSUE TITLE
IShellItem::GetDisplayName bindings

### ISSUE THREAD

**Issue by rednoah** [dbwiddis/IFileDialogImp#1]
So far so good. I've got a test app that successfully opens the system file dialog:
https://github.com/rednoah/IFileDialogImp/blob/main/test/Main.java#L45

I got stuck at getting the selected file paths, which seems to require additional mappings:
* https://docs.microsoft.com/en-us/windows/win32/api/shobjidl_core/nf-shobjidl_core-ifileopendialog-getresults
* https://docs.microsoft.com/en-us/windows/win32/api/shobjidl_core/nn-shobjidl_core-ishellitemarray
* https://docs.microsoft.com/en-us/windows/win32/api/shobjidl_core/nf-shobjidl_core-ishellitem-getdisplayname

**Comment by dbwiddis** [comment]
I'll add the `IShellItemArray` mappings shortly.

**Comment by dbwiddis** [comment]
Done in c4827781697d3f57ce75c36ccc8832626153c2ed

**Comment by dbwiddis** [comment]
Actually only did the array version let me add the `IShellItem` also...

**Comment by dbwiddis** [comment]
Done in 4a456d735ddbf86240055f519cd8e8289fcbea17

Still room to add `IBindCtx`, `IFileDialogEvents`, `IShellItemFilter`, `IPropertyStore`, `IPropertyDescriptionList`, and `IFileOperationProgressSink` if any of those are needed, just let me know.

**Comment by dbwiddis** [comment]
You're using the old/deprecated COM initializer. I learned the hard way that you need to track the return value to know whether you successfully initiaized first time or initialized with the same mode (in which case you have to uninit) or if it failed due to wrong mode (in which case don't uninit or you will kill other programs!).  The below lines of code are hard-earned. ;)

For [this line](https://github.com/rednoah/IFileDialogImp/blob/096281c0c6ed152a1241d8fd668ffcd6df950ddb/test/Main.java#L47) I suggest replacing `Ole32.INSTANCE.CoInitialize(null);` with:
```
    boolean initCOM() {
        boolean comInit = false;
        comInit = initCOM(Ole32.COINIT_MULTITHREADED);
        if (!comInit) {
            comInit = initCOM(Ole32.COINIT_APARTMENTTHREADED);
        }
        return comInit;
    }

    boolean initCOM(int coInitThreading) {
        WinNT.HRESULT hres = Ole32.INSTANCE.CoInitializeEx(null, coInitThreading);
        switch (hres.intValue()) {
        // Successful local initialization (S_OK) or was already initialized
        // (S_FALSE) but still needs uninit
        case COMUtils.S_OK:
        case COMUtils.S_FALSE:
            return true;
        // COM was already initialized with a different threading model
        case WinError.RPC_E_CHANGED_MODE:
            return false;
        // Any other results is impossible
        default:
            throw new COMException("Failed to initialize COM library.", hres);
        }
    }
```
and add
```
    void unInitCOM() {
        Ole32.INSTANCE.CoUninitialize();
    }
```
Store that `initCOM()` boolean to know whether you need to uninit later.
```
        boolean comInit = false;
        try {
            comInit = initCOM();
            // Do stuff with COM
        } catch (COMException e) {
            // Handle exceptions
        }
        if (comInit) {
            unInitCOM();
        }
```

**Comment by dbwiddis** [comment]
For [this line (and the next two)](https://github.com/rednoah/IFileDialogImp/blob/096281c0c6ed152a1241d8fd668ffcd6df950ddb/test/Main.java#L49) you could just call `FileOpenDialog vFileDialog = FileOpenDialog.create();` which you can implement in the `FileOpenDialog` class:
```
        public static FileOpenDialog create() {
            PointerByReference pbr = new PointerByReference();

            HRESULT hres = Ole32.INSTANCE.CoCreateInstance(CLSID_FILEOPENDIALOG, null, WTypes.CLSCTX_ALL,
                    IID_IFILEOPENDIALOG, pbr);
            if (COMUtils.FAILED(hres)) {
                return null;
            }

            return new FileOpenDialog(pbr.getValue());
        }
```

**Comment by rednoah** [comment]
Thanks for the additions and suggestions for using COM correctly. I'll do some testings later tonight and report back.

**Comment by rednoah** [comment]
Yep, works now. Had to modify `GetDisplayName` to use `PointerByReference` though to get the file path value out:
https://github.com/rednoah/IFileDialogImp/blob/main/test/IFileDialogUtil.java#L66

**Comment by dbwiddis** [comment]
Good point on the WString pointer.  I think there are a few other WStrings hanging around, I'll fix those.

**Comment by dbwiddis** [comment]
Fixed in 409ab618f9d82635802d7254246575272d942835

**Comment by dbwiddis** [comment]
Also re: `// Any other results is impossible` to clarify that's "expected results". There are other potential error codes for major issues like out-of-memory, etc. which is why the exception has to be handled.

**Comment by dbwiddis** [comment]
Added IFileOperation mapping in 43c2334c59a01a9ba8c6befab6005aa089daaa58

**Comment by rednoah** [comment]
`IFileOperation` works perfectly in my test app. I just had to make it `public` for other packages. 💯
